src/appleseed.studio/mainwindow/rendering/lightpathswidget.cpp

```diff
@@ -160,149 +193,539 @@ void LightPathsWidget::slot_synchronize_camera()
     }
 }
 
-void LightPathsWidget::paintGL()
+void LightPathsWidget::initializeGL()
 {
-    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+    // If there was already previous data, clean up
+    LightPathsWidget::cleanup_gl_data();
+
+    const auto qgl_format = format();
+
+    if (!initializeOpenGLFunctions())
+    {
+        const int major_version = qgl_format.majorVersion();
+        const int minor_version = qgl_format.minorVersion();
+        RENDERER_LOG_ERROR(
+            "opengl: could not load required gl functions. loaded version %d.%d, required version 3.3",
+            major_version,
+            minor_version);
+        m_gl_initialized = false;
+        return;
+    }
+
+    if (qgl_format.samples() != static_cast<int>(m_gl_msaa_samples))
+    {
+        RENDERER_LOG_WARNING(
+            "opengl: could not enable requested msaa samples. msaa samples enabled: %d",
+            qgl_format.samples());
+    }
+
+    bool manual_srgb_conversion = false;
+    if (qgl_format.colorSpace() != QSurfaceFormat::sRGBColorSpace)
+    {
+        manual_srgb_conversion = true;
+        RENDERER_LOG_WARNING(
+            "opengl: srgb framebuffer extensions not supported, using slower manual conversion.");
+    }
+
+    glEnable(GL_DEPTH_TEST);
+
+    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
+
+    const QString scene_frag_shader_path = manual_srgb_conversion
+        ? ":/shaders/scene_manual_srgb.frag"
+        : ":/shaders/scene.frag";
+
+    create_shader_program(
+        m_scene_shader_program,
+        load_gl_shader(":/shaders/scene.vert"),
+        load_gl_shader(scene_frag_shader_path));
+
+    create_shader_program(
+        m_light_paths_shader_program,
+        load_gl_shader(":/shaders/lightpaths.vert"),
+        load_gl_shader(":/shaders/lightpaths.frag"));
+
+    m_scene_view_mat_location = glGetUniformLocation(m_scene_shader_program, "u_View");
+    m_scene_proj_mat_location = glGetUniformLocation(m_scene_shader_program, "u_Proj");
+    m_scene_camera_pos_location = glGetUniformLocation(m_scene_shader_program, "u_CameraPos");
+
+    m_light_paths_view_mat_location = glGetUniformLocation(m_light_paths_shader_program, "u_View");
+    m_light_paths_proj_mat_location = glGetUniformLocation(m_light_paths_shader_program, "u_Proj");
+
+    const float near_val = 0.01f;
+    const float far_val = 1000.0f;
 
     const auto& rc = m_camera.get_rasterization_camera();
 
-    glMatrixMode(GL_PROJECTION);
-    glLoadIdentity();
+    const float fy = tan(rc.m_hfov / rc.m_aspect_ratio * 0.5) * near_val;
+    const float fx = fy * rc.m_aspect_ratio;
+
+    const float shift_x = rc.m_shift_x * 2.0 * fx;
+    const float shift_y = rc.m_shift_y * 2.0 * fy;
+
+    const float left   = -fx + shift_x;
+    const float right  =  fx + shift_x;
+    const float top    = -fy + shift_y;
+    const float bottom =  fy + shift_y;
+
+    // Reference: https://www.cs.rit.edu/usr/local/pub/wrc/graphics/doc/opengl/books/blue/glFrustum.html
```

Comment:
I was tempting to suggest adding a new `make_frustum()` static method to `Matrix<T, 4>` but this code seems rather OpenGL-specific. That said, maybe by reversing top and bottom when invoking it and transposing its result, we can make `make_frustum()` generic and compliant with appleseed's conventions?

Reply:
It worked :)